### PR TITLE
Reimplement spectral_norm using new parametrization functionality

### LINKED ISSUE #51800: spectral_norm not properly initialized if used with eval first
According to the documentation, `nn.utils.spectral_norm` first performs some initialization actions the first time that `forward` is called. However, if `forward` is first called in `eval` mode, this initialization does not occur. The combination of the lack of initialization and how `spectral_norm` layers are evaluated in forward passes can result in layer outputs with values several orders of magnitude larger than the inputs. This scenario of a model being used in `eval` mode before `train` can arise naturally with some GAN training procedures, or when trying to assess initializations and their effect on activation variance across layers.

There is a workaround for this problem: perform a single forward pass in `train` mode first. However, unnoticed, it can silently cause training problems or be hard to diagnose. To avoid this unnecessary and potentially harmful complication, spectral_norm should be modified to either perform initialization immediately upon being applied to a layer, or to perform it in the first forward pass even if it is not in `train` mode.

Sample code to reproduce:
```py
    class SpectralNormTestModel(nn.Module):
        
        def __init__(self, use_spectral_norm: bool):
            super(SpectralNormTestModel, self).__init__()
            feature_count = 1000

            def init_(layer: nn.Linear):
                nn.init.orthogonal_(layer.weight.data, 1.372)
                nn.init.zeros_(layer.bias.data)

            self.fc1 = nn.Linear(feature_count, feature_count)
            init_(self.fc1)
            if use_spectral_norm:
                self.fc1 = nn.utils.spectral_norm(self.fc1)
            self.act1 = nn.PReLU(feature_count)

            self.fc2 = nn.Linear(feature_count, feature_count)
            init_(self.fc2)
            if use_spectral_norm:
                self.fc2 = nn.utils.spectral_norm(self.fc2)
            self.act2 = nn.PReLU(feature_count)

            self.out = nn.Linear(feature_count, 1)
            init_(self.out)

        def forward(self, x: torch.Tensor) -> torch.Tensor:
            return self.out(self.act2(self.fc2(self.act1(self.fc1(x)))))
    
    
    def spectral_norm_init_test(use_spectral_norm: bool):
        model = SpectralNormTestModel(use_spectral_norm).cuda().eval()
        x = torch.randn(10000, 1000).cuda()
        y = model(x)
        print(f"mean absolute output: {y.abs().mean()}")
```
When run without using spectral norm, the mean absolute output stays between around 0.9 and 1.8. With spectral norm, it commonly varies between around 100 and 100000. 

I am using torch version 1.7.1+cu110 in Python 3.8.5 on Windows 10.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @albanD @mruberry

### PR DESCRIPTION
**Pull request by soulitzer**
Adds a new file under `torch/nn/utils/parametrizations.py` which should contain all the parametrization implementations

For spectral_norm we add the `SpectralNorm` module which can be registered using `torch.nn.utils.parametrize.register_parametrization` or using a wrapper: `spectral_norm`, the same API the old implementation provided.

Most of the logic is borrowed from the old implementation:
 - Just like the old implementation, there should be cases when retrieving the weight should perform another power iteration (thus updating the weight) and cases where it shouldn't. For example in eval mode `self.training=True`, we do not perform power iteration.

There are also some differences/difficulties with the new implementation:
 - Using new parametrization functionality as-is there doesn't seem to be a good way to tell whether a 'forward' call was the result of parametrizations are unregistered (and leave_parametrizations=True) or when the injected property's getter was invoked. The issue is that we want perform power iteration in the latter case but not the former, but we don't have this control as-is. So, in this PR I modified the parametrization functionality to change the module to eval mode before triggering their forward call
 - Updates the vectors based on weight on initialization to fix https://github.com/pytorch/pytorch/issues/51800 (this avoids silently update weights in eval mode). This also means that we perform twice any many power iterations by the first forward.
 - right_inverse is just the identity for now, but maybe it should assert that the passed value already satisfies the constraints
 - So far, all the old spectral_norm tests have been cloned, but maybe we don't need so much testing now that the core functionality is already well tested

### CODE REVIEW

File: torch/nn/utils/parametrizations.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+            calculating norms
+        dim (int, optional): dimension corresponding to number of outputs,
+            the default is ``0``, except for modules that are instances of
+            ConvTranspose{1,2,3}d, when it is ``1``
+
+    Returns:
+        The original module with a new parametrization registered to the specified
+        weight
+
+    .. note::
```
Comment (Lezcano) [inline torch/nn/utils/parametrizations.py:156]: Nice!

**Review by lezcano — COMMENTED** [review]
This looks very good! I left a few comments with some simplifications that could be made.

It's very nice to see how easy it is to read through the code now when compared to the version without parametrisations really.

**Edit.** On the differences:
- I have thought a bit about this, and I'd say that this is fine. @albanD can you have a look at it?
- I would not update the weight in eval mode, as that could really bite people
- I think setting the right_inverse to the identity is fine
- Can we perhaps refactor the tests or test the functionality of the two at the same time?

File: torch/nn/utils/parametrizations.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        This function is implemented using new parametrization functionality
```
Comment (Lezcano) [inline torch/nn/utils/parametrizations.py:157]: Nit. using new -> using the new

File: torch/nn/utils/spectral_norm.py [hunk 3]
```diff
@@ -251,6 +251,14 @@ def spectral_norm(module: T_module,
     Returns:
         The original module with the spectral norm hook
 
+    .. note::
+        This function has been reimplemented as
+        :func:`torch.nn.utils.parametrizations.spectral_norm` using new
+        parametrization functionality in
+        :func:`torch.nn.utils.parametrize.register_parametrization`. Please use
+        the newer version. This function will be deprecated in a future version
+        of PyTorch.
```
[the remark this answers is not in the document]
Comment (albanD) [reply torch/nn/utils/spectral_norm.py:260]: We don't want to do this for 1.9 as parametrization is very early and other things like weight_norm will not be ported yet.
We will add the deprecation warning after the branch cut.

File: torch/nn/utils/parametrizations.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+                                         dim=0, eps=self.eps, out=self.v)   # type: ignore[has-type]
+                    self.u = F.normalize(torch.mv(weight_mat, self.v),
+                                         dim=0, eps=self.eps, out=self.u)   # type: ignore[has-type]
+                # See above on why we need to clone
+                self.u = self.u.clone(memory_format=torch.contiguous_format)
+                self.v = self.v.clone(memory_format=torch.contiguous_format)
+        else:
+            # https://github.com/pytorch/pytorch/issues/51800
+            if not self.buffers_initialized:
+                warnings.warn("SpectralNorm's `forward` called for the first time, but "
```
Comment (albanD) [inline torch/nn/utils/parametrizations.py:94]: Why do you need this? You are initializing v and u based on the given weight in the init no?

File: torch/nn/utils/parametrize.py [hunk 5]
```diff
@@ -341,9 +341,17 @@ def remove_parametrizations(
...
     # Fetch the original tensor
-    original = module.parametrizations[tensor_name].original  # type: ignore[index, union-attr]
+    parametrizations = module.parametrizations[tensor_name]  # type: ignore[index, union-attr]
+    original = parametrizations.original  # type: ignore[union-attr]
     if leave_parametrized:
-        t = getattr(module, tensor_name)
+        # Call getattr in eval mode so that parametrizations that
+        # compute weight iteratively won't perform another iteration
+        training = parametrizations.training   # type: ignore[union-attr]
+        parametrizations.eval()  # type: ignore[union-attr]
```
Comment (albanD) [inline torch/nn/utils/parametrize.py:350]: That looks like a pretty big change and that can have pretty big impact. I don't think you want to change that.

File: torch/nn/utils/parametrizations.py (hunk 4 above)
Comment (Lezcano) [reply torch/nn/utils/parametrizations.py:94]: I thought about this the other day and I forgot to leave a comment.
We have access to the `weight` during `__init__`, so I believe it'd be good to just initialize `u` and `v` correctly in the `__init__` and not have this else nor the `buffers_initialized` attribute.

Comment (soulitzer) [reply torch/nn/utils/parametrizations.py]: Fixed.

File: torch/nn/utils/parametrize.py (hunk 5 above)
Comment (soulitzer) [reply torch/nn/utils/parametrize.py:350]: Hmm, aren't the parametrizations about to be removed? We also restore it to its original state since the user may be holding a pointer as @Lezcano mentioned. 

One issue I see with the current PR is that individual parametrizations may have different values of eval vs train. If that is the only issue, we could save and restore those values individually over a loop

Another way is to deepcopy the parametrizations and invoke its forward directly instead of doing getattr

Comment (albanD) [reply torch/nn/utils/parametrize.py:350]: What I mean is that this can be surprising for the user that the removal of the reparametrization happens in eval mode even though they didn't ask for it.

I would feel better without this code and a Note in the spectral_norm doc: you should put the Module in eval() mode before removing the reparametrization if you don't want to do an extra iteration.

Comment (soulitzer) [reply torch/nn/utils/parametrize.py]: Removed the code and added the note.
Side note:  if we register a parametrization on a module in eval mode, its parametrization module would actually still be training mode, and this may cause it to perform updates silently. To be consistent with the behavior of the old
spectral norm, maybe newly registered parametrizations should default to the training mode of its host module.

```
pre_remove_out = m(input)
m.eval() # this works as expected because SpectralNorm is a submodule
m = torch.nn.utils.parametrize.remove_parametrizations(m, 'weight')
self.assertEqual(m(input), pre_remove_out)

torch.nn.utils.parametrizations.spectral_norm(m)
pre_remove_out = m(input)
self.assertTrue(m.parametrizations.weight[0].training) # m is actually still in eval mode, but newly registered
                                                                                     # parametrization is in training mode
m = torch.nn.utils.parametrize.remove_parametrizations (m, 'weight')
self.assertNotEqual(m(input), pre_remove_out)
```

Comment (albanD) [reply torch/nn/utils/parametrize.py]: training mode of its host Module sounds good to me, what do you think @Lezcano

Comment (soulitzer) [reply torch/nn/utils/parametrize.py]: Update: I've tentatively added this feature in this PR

Comment (Lezcano) [reply torch/nn/utils/parametrize.py]: I think training mode of the host is reasonable.

File: torch/nn/utils/parametrize.py [hunk 6]
```diff
@@ -341,9 +341,15 @@ def remove_parametrizations(
...
 
     # Fetch the original tensor
-    original = module.parametrizations[tensor_name].original  # type: ignore[index, union-attr]
+    parametrizations = module.parametrizations[tensor_name]  # type: ignore[index, union-attr]
+    original = parametrizations.original  # type: ignore[union-attr]
     if leave_parametrized:
+        # Call getattr in eval mode so that parametrizations that
+        # compute weight iteratively won't perform another iteration
+        for m in parametrizations:
+            m.training = False
```
[the remark this answers is not in the document]
Comment (Lezcano) [reply torch/nn/utils/parametrize.py]: @albanD @soulitzer even if we do not doi the training bit, I think it's still good to do the `torch.no_grad` here, as we do not really need the gradients. I know this is not related to this PR, but it's just something I realised that I forgot to add to the main parametrizations PR. If you don't want to add it here I can add it in a separate PR.

File: torch/nn/utils/parametrizations.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+        #          is already on correct device; and it makes sure that the
+        #          parallelized module is already on `device[0]`.
+        #       2. If the out tensor in `out=` kwarg has correct shape, it will
+        #          just fill in the values.
+        #     Therefore, since the same power iteration is performed on all
+        #     devices, simply updating the tensors in-place will make sure that
+        #     the module replica on `device[0]` will update the _u vector on the
+        #     parallized module (by shared storage).
+        #
+        #    However, after we update `u` and `v` in-place, we need to **clone**
```
Comment (albanD) [inline torch/nn/utils/parametrizations.py:64]: I'm not really buying this. They are modified inplace with an `out=` function which does not support autograd. So I'm really surer how that can happen.
But I can we can check that out later.

**Review by albanD — APPROVED** [review]
Beyond the name update I think this is good to go.
I would have some more questions about the implementation but that's taken from the existing one so that can wait I think.

File: torch/nn/utils/parametrize.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (soulitzer) [reply torch/nn/utils/parametrize.py]: Added!

File: torch/nn/utils/parametrizations.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
+import torch
+from ..utils import parametrize
+from ..modules import Module
+from .. import functional as F
+
+from typing import Optional
+
+class SpectralNorm(Module):
```
[the remark this answers is not in the document]
Comment (soulitzer) [reply torch/nn/utils/parametrizations.py]: Making it internal

File: torch/nn/utils/parametrizations.py (hunk 7 above)
Comment (albanD) [reply torch/nn/utils/parametrizations.py:64]: Ho that whole function is in no_grad mode. So no gradients here.
And indeed these will be saved in the forward so the clone is good.

Comment (Lezcano) [reply torch/nn/utils/parametrizations.py:64]: Oh, now that's interesting. I wonder whether it's possible to approximate the derivative with respect to the largest singular values of a matrix without computing the whole SVD... I'll think about this.

Comment (Lezcano) [reply torch/nn/utils/parametrizations.py:64]: Nevermind. It turns out that differentiating this expression with `u` and `v` fixed actually gives the correct gradients for the expression `A / \norm{A}_2`

**Comment by AdilZouitine** [comment]
Great Pull request ! will this new spectral norm be available on pytorch 1.9?
